### ISSUE TITLE
Enable Marantz SR5007 Internet Radio Favorites 

### ISSUE THREAD

**Issue by LePremier** [foxriver76/ioBroker.denon#102]
Hi,

I have the problem that I am not able to select (set) internet radio favorites via settings.savePreset (settings.loadPreset). Also the info.onlinePresets JASON string is empty / not changed when editing the fields in ioBroker*. 

However it is in basically possible to administrate the favorites remotely via the standard DENON Remote App on iOS.

It would be absolutely brilliant if you could add that feature for the Marantz SR5007 to the ioBroker.denon adapter. 

Many thanks,

Jonathan

* used version 1.9.1 of Denon AVR adapter in ioBroker admin version 4.1.11.; I tried editing settings.savePreset field in different ways: 
1. just entering a number: 1
2. entering the number with "": "1"
3. entering in format: {"id":1,"channel":"ORF FM4"}
4. entering in format: "id":1,"channel":"ORF FM4"

**Comment by foxriver76** [comment]
Just entering the number should be all as stated by the documentation, just tested it, and works at least on my AVR X1200W. Of course, you have to listen actively to "Internet Radio" to save the current channel. It can be, that Marantz has different Commands on your receiver, unfortunately, I am not able to find a document for your specific Receiver and the general Marantz Protocols do not include commands for the internet radios presets. If you find a valid API description or get the working commands, I would love to implement them.

Edit: I found this [document](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwio2qDM1bvuAhUa7eAKHVbbDloQFjAAegQIAxAC&url=https%3A%2F%2Fwww.marantz.com%2F-%2Fmedia%2Ffiles%2Fdocumentmaster%2Fmarantzna%2Fus%2Fmarantz-2014-nr-series---sr-series-rs232-ip-protocol.xls&usg=AOvVaw0JZPwfyROvKYDxO3wT59Od) which states that it should work exactly the same to save favorites like it is implemented, but your AVR is not listed, so not 100 % sure. 

So please try again, set adapter instance to loglevel debug, listen to an Internet Radio channel and try to save it e.g. by adding 1 in `savePreset` and post the log here.

**Comment by LePremier** [comment]
Thanks for the quick response and offer to support!

I switched the adapter logging to debug mode, based on your great YouTube video :) Afterwards I set 4 favorites by simply tipping a number in the settings.savePreset and and confirming with enter button, while listening to the corresponding station station. I started this procedure with id:0.

Doing so I got a bit of a strange JSON. It includes entries for two stations (BBC Radio 6 at id:1 and ORF FM4 at id:2). However it doesn't have any entries for the two other stations (ByteFM and egoFM).

Afterwards I tired to switch to the saved stations by entering the corresponding number into the settings.loadPreset field. This is possible for BBC Radio 6 and ORF FM4.

Here are the log-file extract and a copy of the JASON file:

---
LOG
---

[iobroker.2021-01-28.log.txt](https://github.com/foxriver76/ioBroker.denon/files/5886251/iobroker.2021-01-28.log.txt)
-> please open as html

---
JSON
---

[{"id":0,"channel":""},{"id":1,"channel":"BBC Radio 6 Music Li"},{"id":2,"channel":"ORF FM4 Live"},{"id":3,"channel":""},{"id":4,"channel":""},{"id":5,"channel":""},{"id":6,"channel":""},{"id":7,"channel":""},{"id":8,"channel":""},{"id":9,"channel":""},{"id":10,"channel":""},{"id":11,"channel":""},{"id":12,"channel":""},{"id":13,"channel":""},{"id":14,"channel":""},{"id":15,"channel":""},{"id":16,"channel":""},{"id":17,"channel":""},{"id":18,"channel":""},{"id":19,"channel":""},{"id":20,"channel":""},{"id":21,"channel":""},{"id":22,"channel":""},{"id":23,"channel":""},{"id":24,"channel":""},{"id":25,"channel":""},{"id":26,"channel":""},{"id":27,"channel":""},{"id":28,"channel":""},{"id":29,"channel":""},{"id":30,"channel":""},{"id":31,"channel":""},{"id":32,"channel":""},{"id":33,"channel":""},{"id":34,"channel":""},{"id":35,"channel":""},{"id":36,"channel":""},{"id":37,"channel":""},{"id":38,"channel":""},{"id":39,"channel":""},{"id":40,"channel":""},{"id":41,"channel":""},{"id":42,"channel":""},{"id":43,"channel":""},{"id":44,"channel":""},{"id":45,"channel":""},{"id":46,"channel":""},{"id":47,"channel":""},{"id":48,"channel":""},{"id":49,"channel":""},{"id":50,"channel":""},{"id":51,"channel":""},{"id":52,"channel":""},{"id":53,"channel":""},{"id":54,"channel":""},{"id":55,"channel":""}]

**Comment by foxriver76** [comment]
Hm okay, then it basically seems to work, as the JSON states it has only saved `BBC Radio 6` to preset 1 and `ORF FM4` to preseet 2. Thus, only these two are working. According to the log, the behavior looks good, so I am quite unsure why some channels are not saved. 

As stupid as it sounds for now I would just try again, to save the preset if you see it hasn't been saved to `info.onlinePresets`. In the log I also cannot really detect if the AVR currently is playing a station, because we just can say -> whatever station is playing, save it to `01` then we just send this `01` and we are requesting the whole list again, and with this list we will update the `onlinePresets` state.

However, I've seen that the adapter is doing unnecessary work, because it hasn't returned after handling the NSH commands which lead to `2021-01-28 09:56:23.198 - debug: denon.0 (355842) [INFO] <== Unhandled command NSH` messages. This is fixed with 1.10.2

**Comment by LePremier** [comment]
I've tried to bypass the issue by simply repeating the operations again and again. But nothing happens. I.e. no new stations are stored in the JASON. I did this for all various radio stations, while I was listening to them :(

I am not really knowing the meanings of the log entries. But it looks kind of strange to me that I try to add a certain station but in the following entries this station isn't mentioned at all. Instead of that there are entries corresponding to BBC Radio and FM4 (i.e. the stations that are already in the JASON)

For example 
- I listened to "Radio Swiss Jazz"
- I tried to add it by setting "State Change - ID: settings.savePreset; State: 3"
- I get entries like "Incoming data: NSH01BBC Radio 6 Music Li" , "Command to handle is NSHBBCRadioMusicLi", "Unhandled command NSHBBCRadioMusicLi", "Incoming data: NSH02ORF FM4 Live", "Command to handle is NSHORFFMLive", "Unhandled command NSHORFFMLive"
- Afterwards I get a long list of entries in the style ""Incoming data: NSHxx", ""Command to handle is Command to handle is NSH", "Unhandled command NSH" -> I think that's what you already said.

This is reproducible for Radio Swiss Jazz as well as any other station.

Maybe it helps - or is it all "usual business"?

**Comment by foxriver76** [comment]
> I get entries like "Incoming data: NSH01BBC Radio 6 Music Li" , "Command to handle is NSHBBCRadioMusicLi", "Unhandled command NSHBBCRadioMusicLi", "Incoming data: NSH02ORF FM4 Live", "Command to handle is NSHORFFMLive", "Unhandled command NSHORFFMLive"

Yes, NSH**01**BBC Radio 6 Music Li, means that the favorite which is located on preset 1 is `BBC Radio 6 Music Li` and it will send this for all favorites, also if there are no favorites at a preset, the AVR will just send us `NSH05` without any name behind it. Because the AVR won't tell us by itself that a preset has successfully changed, we have to request the whole list of presets after "trying to change" a preset.

I can try to replicate it with the mentioned station "Radio Swiss Jazz" if I am able to find it.

**Comment by foxriver76** [comment]
Hm.. okay, noticed that you now need a vTuner subscription to search for new stations... so the only channels are left to select are "recently played", so I am not really able to search for the specific station.

**Comment by LePremier** [comment]
Yeah, I had the same issue few weeks ago. I am happy to sponsor a one year subscription, if that might help? ;)

**Comment by LePremier** [comment]
I just sent a small donation via the github sponsoring program. Even if it won't lead to a fix of the current issue I am happy to support the great efforts with that you already have contributed to the iobroker community.

**Comment by foxriver76** [comment]
Thank you, I now have a subscription and will test it.

**Comment by foxriver76** [comment]
Hm before setting `savePreset` to `2`
```
[{"id":0,"channel":"Antenne Bayern"},{"id":1,"channel":"Antenne Bayern"},{"id":2,"channel":"Radio Swiss Jazz"},{"id":3,"channel":"Antenne Bayern"},{"id":4,"channel":""},{"id":5,"channel":""},{"id":6,"channel":""},{"id":7,"channel":"104.6 RTL Best of Bl"},{"id":8,"channel":""},{"id":9,"channel":""},{"id":10,"channel":""},{"id":11,"channel":""},{"id":12,"channel":""},{"id":13,"channel":""},{"id":14,"channel":""},{"id":15,"channel":"104.6 RTL Best of Bl"},{"id":16,"channel":""},{"id":17,"channel":""},{"id":18,"channel":""},{"id":19,"channel":""},{"id":20,"channel":""},{"id":21,"channel":""},{"id":22,"channel":""},{"id":23,"channel":""},{"id":24,"channel":""},{"id":25,"channel":""},{"id":26,"channel":""},{"id":27,"channel":""},{"id":28,"channel":""},{"id":29,"channel":""},{"id":30,"channel":"Antenne Bayern"},{"id":31,"channel":""},{"id":32,"channel":""},{"id":33,"channel":""},{"id":34,"channel":""},{"id":35,"channel":""}]
```

After:
```
[{"id":0,"channel":"Antenne Bayern"},{"id":1,"channel":"Antenne Bayern"},{"id":2,"channel":"Radio Swiss Jazz"},{"id":3,"channel":"Antenne Bayern"},{"id":4,"channel":""},{"id":5,"channel":""},{"id":6,"channel":""},{"id":7,"channel":"104.6 RTL Best of Bl"},{"id":8,"channel":""},{"id":9,"channel":""},{"id":10,"channel":""},{"id":11,"channel":""},{"id":12,"channel":""},{"id":13,"channel":""},{"id":14,"channel":""},{"id":15,"channel":"104.6 RTL Best of Bl"},{"id":16,"channel":""},{"id":17,"channel":""},{"id":18,"channel":""},{"id":19,"channel":""},{"id":20,"channel":""},{"id":21,"channel":""},{"id":22,"channel":""},{"id":23,"channel":""},{"id":24,"channel":""},{"id":25,"channel":""},{"id":26,"channel":""},{"id":27,"channel":""},{"id":28,"channel":""},{"id":29,"channel":""},{"id":30,"channel":"Antenne Bayern"},{"id":31,"channel":""},{"id":32,"channel":""},{"id":33,"channel":""},{"id":34,"channel":""},{"id":35,"channel":""}]
```

It so strange that it worked for like two stations for you and now it doesn't. But the two stations inside your `onlinePresets` weren't there before you saved them today, I guess?

**Comment by LePremier** [comment]
> But the two stations inside your onlinePresets weren't there before you saved them today, I guess?

... i think that is indeed the case for FM4. BBC Radio 6 was added today (at least I think so)... 

I think, just editing the JASON wouldn't be a valid workaround?

**Comment by foxriver76** [comment]
You could also try to send the command manually, but according to the log the commands look good as mentioned before.

Try `settings.expertCommand` while listening to your station, then set `NSC03` to `settings.expertCommand`.



> I think, just editing the JASON wouldn't be a valid workaround?

Won't do anything the presets are really stored on AVR side. You could also increase the intervals in the adapter config, maybe we are sending too much too fast and commands are getting discarded on AVR side, but I don't think so.

**Comment by LePremier** [comment]
Really curios: I did dozens of tries and now ByteFM was added to channel 4. Trying again with other stations again doesn't work (like it was also with ByteFM in former tries). No matter using the settings.savePreset nor the settings.expertCommand.

It feels like storing only works under certain conditions / at certain time-slots.

Is the adapter using the same API like the Denon app? I am able to add / remove stations there. Afterwards they are listed as favorites in the receivers menu. Also it is possible to switch channel remotely by the app. But I guess this favorites aren't the same like the onlinePresets we are talking about in the context of the ioBroker adapter?! Also at the receiver itself I think there is no way to directly browse the onlinePresets. Is this the same for your Denon device?

**Comment by foxriver76** [comment]
AFAIK the smartphone app uses another API, because the AVR only allows one telnet connection, and I am able to control my AVR via the app while the adapter is connected.

Have you tried increasing the time between pollings/commands in the adapter settings?

**Comment by foxriver76** [comment]
How many favorites can be specified in the app? Maybe it's also possible to control those too via telnet API. Are you using the official DENON app or something special for Marantz?

**Comment by LePremier** [comment]
> Have you tried increasing the time between pollings/commands in the adapter settings?

Not yet. Will try asap.

I am using the official Denon App. I haven't seen / reached a maximum level of favourites there. Currently I have 15 stations in the list.

> Maybe it's also possible to control those too via telnet API

That would be great!

**Comment by foxriver76** [comment]
![Uploading E68619A8-59B7-4C8A-AFCE-1FC5B35E8108.png…]()

**Comment by foxriver76** [comment]
![FE806CEB-354C-4A56-A94D-A43477F4AF85](https://user-images.githubusercontent.com/25484431/106249339-8aa5ff00-6212-11eb-9038-f88738d3e91c.png)

**Comment by LePremier** [comment]
Ok, I played around a little with the delay times between the commands (Adapter Configuration -> "Erweiterte Einstellungen"). 

Thereby **50ms** seems to be good value for my setup with the **Marantz SR5007**. It led to a reasonable reduction of "<== Unhandled command NSH" log entries. 

But more important: with this setup I worked out to **add several radio stations** to the onlinePresets JSON. This was possible with the following steps (others might work as well):

1. Open Denon App (in my case on iOS)
2. Add the radio stations to apps favorites list
3. Select the radio station you wish to add to onlinePresets JSON
4. Wait until the Marantz receiver has started the playback of the station
5. Edit the field settings.expertCommand with `NSCxx` (thereby `xx` is the id of the channel you want to save the station at in the preset.json
6. Confirm with enter button 

In some cases i was required to repeat step 5. and 6. for several times. But in the end the station was always stored in the JSON.

By repeating steps 3-6 I was able to add all of my stations. 

The fields savePreste and loadPreset have been empty while i did the procedure.

I am not sure if I was just lucky or if this is really reproducible. Therefore I will try the workflow for some more times in the next days and will keep you in the loop. 

On last word regarding the delay times: much higher latencies (e.g. 1000 ms) lead to a lot of additional entries in the log. So maybe this is something that seems to be an important influencing factor. If you have any ideas how I can support you with optimizing it (based on the SR5007), please let me know.

**Comment by foxriver76** [comment]
Okay, this is good news.

I am not sure if this has really an influence or is just a coincidence with your step-by-step procedure.

However, you can try the latest version v1.10.3 which now reduces the load when saving a channel with savePreset + some more optimizations on the performance side.

> Select the radio station you wish to add to onlinePresets JSON
Wait until the Marantz receiver has started the playback of the station

This should of course be the normal procedure for adding a radio station to the onlinePresets, expect that you normally don't need to add the station to your favorites via remote. Of course now save preset is doing the same as NSC expertCommand w/o additional unnecessary things. So steps 5-6 can be safely replaced by using the state.

**Comment by LePremier** [comment]
> I am not sure if this has really an influence or is just a coincidence with your step-by-step procedure

Yeah, I fully agree. Maybe the current constellation of Mars and Venus has more impact on the success than the changes I tried with legacy, etc. :-D But I yet wanted to mention so that someone that have similar issues can as well give it a try. 

> This should of course be the normal procedure

It was in all of my attempts. I just wanted to mention it to give the complete picture of my workflow. 

> So steps 5-6 can be safely replaced by using the state.

Unfortunately it doesn't **seems** to work in my case. I just tried again. Adding via state field does not work after 3 tries. Using settings.expertCommand with NSCxx worked at the 3rd try. Of course this could again be a coincidence, but up so far setting state never worked for me. Here is the corresponding log. Maybe it can offer some more details to you?

[iobroker.2021-01-29-(1924).log.txt](https://github.com/foxriver76/ioBroker.denon/files/5895632/iobroker.2021-01-29-.1924.log.txt)

I am looking forward to try the new version. I think it's not yet available via ioBroker upgrade function?

Many thanks for all of your good support so far!!!

**Comment by foxriver76** [comment]
> I am looking forward to try the new version. I think it's not yet available via ioBroker upgrade function?

Maybe it took some hours to be available in the latest Repository but normally it's really fast. But now it should be available LOL ;-)

 However, I am not really sure how to proceed with this issue.

**Comment by LePremier** [comment]
I did the update. But unfortunately my SR5007 in general seems not working very reliable with the adapter. Another example is the connection status. After switching the off the receiver it is only updated when also restarting the adapter. 

I worked out to get most of the functionality up and running by integrating specific logic to my Blockly/JS Code. 

I am not sure if this is in the end more an error of the receiver than a compatibility issue with your adapter. Therefore for me, it would be fine to close the issue :)
